packages/compass-editor/src/codemirror/validation-autocompleter.ts

```diff
@@ -0,0 +1,176 @@
+import type { CompletionSource } from '@codemirror/autocomplete';
+import { snippetCompletion } from '@codemirror/autocomplete';
+import { completer } from '../autocompleter';
+import type { CompletionResult } from '../autocompleter';
+
+import {
+  resolveTokenAtCursor,
+  getAncestryOfToken,
+  ARRAY_ITEM_REGEX,
+} from './utils';
+
+const rootJsonSchemaCompletion = (from: number, to: number) => ({
+  filter: false,
+  from,
+  to,
+  options: [
+    snippetCompletion('{\n\t$jsonSchema: {\n\t\t${}\n\t}\n}', {
+      label: '$jsonSchema',
+      info: 'json-schema validation object',
+    }),
+  ],
+});
+
+const createCompletions = (
+  completions: CompletionResult[],
+  text: string,
+  from: number,
+  detail?: string
+) => {
+  return {
+    from,
+    filter: false,
+    options: completions
+      .filter((completion) => {
+        return completion.value.toLowerCase().startsWith(text.toLowerCase());
+      })
+      .map((completion) => ({
+        ...completion,
+        label: completion.value,
+        info: completion.description,
+        detail: detail ?? completion.meta,
+      })),
+  };
+};
+
+const isCompletingRequired = (ancestors: string[]) => {
+  // Required is always an array.
+  return (
+    (ancestors[ancestors.length - 1] ?? '').match(ARRAY_ITEM_REGEX) &&
+    ancestors[ancestors.length - 2] === 'required'
+  );
+};
+
+const isCompletingProperties = (ancestors: string[]) => {
+  return ancestors[ancestors.length - 1] === 'properties';
+};
+
+const isCompletingBsonType = (ancestors: string[]) => {
+  // bsonType can either be a string or an array.
+  const immediateParent = ancestors[ancestors.length - 1] ?? '';
+  return (
+    immediateParent === 'bsonType' ||
+    (immediateParent.match(ARRAY_ITEM_REGEX) &&
+      ancestors[ancestors.length - 2] === 'bsonType')
+  );
+};
+
+/**
+ * The $jsonSchema's *required* and *properties* fields are nested,
+ * so we need to get the ancestor to get the correct field name.
+ */
+const getNestingAncestor = (ancestors: string[]): string => {
+  return (
+    ancestors
+      .filter((x) => !['$jsonSchema'].includes(x))
+      .map((x, i, a) => {
+        if (['required', 'properties'].includes(x)) {
+          return a[i - 1];
+        }
+      })
+      .filter(Boolean)
+      // because field names are represented in dot notation
+      .join('.')
+  );
+};
+
+/**
+ * As the $jsonSchema's *required* and *properties* fields are nested,
+ * we need to filter the fields based on the nesting level (ancestor).
+ */
+const filterFieldsByAncestor = (
+  fields: CompletionResult[],
+  ancestor: string
+) => {
+  if (ancestor === '') {
+    return fields;
+  }
+
+  return fields
+    .filter((x) => x.value !== ancestor && x.value.startsWith(ancestor))
+    .map((x) => {
+      const value = x.value.replace(`${ancestor}.`, '');
+      return {
+        ...x,
+        value,
+      };
+    })
+    .filter((x) => x.value);
+};
+
+export const createValidationAutocompleter = (
```

Comment:
Seems like this is not doing any autocomplete that was happening before, in particular for the case where you are using [query validation](https://www.mongodb.com/docs/manual/core/schema-validation/specify-query-expression-rules/#std-label-schema-validation-query-expression) and not jsonSchema. Here's current behavior:

![image](https://user-images.githubusercontent.com/5036933/225871926-7806f1fc-88d7-4410-8fa8-4bc239af71fc.png)

Reply:
From the docs, the root validation object either contains `$jsonSchema` or the query operators and not the document fields, but somehow the current auto-completer suggests that. 

When the user is inside the query operator validation, I'll reuse the queryAutoCompleter to handle that efficiently.

Reply:
Gotcha, wasn't obvious if this was removed this on purpose or not